src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java

```diff
@@ -0,0 +1,22 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api;
+
+import com.cloudbees.plugins.credentials.Credentials;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.HttpClientBuilder;
+
+public abstract class BitbucketAuthenticator<T extends Credentials> {
+
+    public final T credentials;
```

Comment:
On the minus side you are basically wrapping a Credentials... please don't do that, extract what you need from the Credential and move on, that way the `BitbucketAuthenticator` can be used from remote agents rather than only from the master JVM

Reply:
Yes that means you need a subclass for each type of credential that is supported... but you needed that anyway. Also... just say no to public fields